File: src/lib/knowledge.ts
// Clients for the Knowledge screen.
//
// The screen reads from two services, and which one owns what is the whole
// arrangement in miniature:
//
//   hostagent  — the graph the user authors (organizations, projects, groups,
//                relations). Persisted, editable while the indexer is down.
//   ragindex   — what the index actually contains: where each source sits in
//                embedding space and which sources are near it.
//
// Nothing here goes through the gateway. These are host-facing management
// routes; the gateway forwards only /search and /source to the indexer,
// precisely because these enumerate every source regardless of group. The two
// it does forward are both filtered by the caller's granted groups — one finds
// a source, the other follows it — while these answer about the whole index.

const HOST = import.meta.env.VITE_HOSTAGENT_URL ?? "http://127.0.0.1:8788";
const RAG = import.meta.env.VITE_RAG_URL ?? "http://127.0.0.1:8790";

async function call<T>(base: string, path: string, init?: RequestInit): Promise<T> {
  const res = await fetch(base + path, {
    ...init,
    headers: { "Content-Type": "application/json", ...(init?.headers ?? {}) },
  });
  if (!res.ok) {
    const body = await res.json().catch(() => ({}));
    throw new Error((body as { error?: string }).error ?? `HTTP ${res.status}`);
  }
  return res.json() as Promise<T>;
}

// --- the authored graph ---

export interface KnowledgeOrg {
  id: string;
  name: string;
}

export interface KnowledgeProject {
  id: string;
  name: string;
  orgId: string;
}

export interface KnowledgeGroup {
  id: string;
  name: string;
  color: string;
  owner: string;
  description: string;
  /** projects this group serves; a group may serve several. */
  projects: string[];
  /** indexed sources this group contains — its retrieval permission set. */
  sources: string[];
}

/** The link kinds the graph can draw, strongest claim first.
 *
 *  Ordering is for reading, not for permission: none of these widen a scope.
 *  The host agent decides which types exist; this list only has to agree. */
export const RELATION_TYPES = [
  "requires",
  "derives-from",
  "same-as",
  "references",
  "supersedes",
  "conflicts-with",
] as const;

export type RelationType = (typeof RELATION_TYPES)[number];

/** Colour and dash per relation type, so the canvas and the inspector agree.
 *
 *  Solid for the edges that assert a positive relationship, dashed for the ones
 *  that qualify or warn. None of them carry permission — a scope is the whole of
 *  what a run may reach — so the styling says what the edge means and nothing
 *  about what it grants. */
export const RELATION_STYLE: Record<RelationType, { color: string; dash: string }> = {
  requires: { color: "var(--ac)", dash: "0" },
  "derives-from": { color: "var(--purple)", dash: "0" },
  "same-as": { color: "var(--cyan)", dash: "0" },
  references: { color: "var(--tx-dim)", dash: "2 3" },
  supersedes: { color: "var(--amber)", dash: "5 3" },
  "conflicts-with": { color: "var(--red)", dash: "5 3" },
};

/** What each relation type asserts. Vocabulary, not permission.
 *
 *  Relations grant nothing: a task's scope is the whole of what its agents may
 *  reach. This exists so the canvas can say what an edge means, which is the
 *  only thing it does. */
export const RELATION_MEANING: Record<RelationType, string> = {
  requires: "needs",
  "derives-from": "provenance",
  "same-as": "identity",
  references: "mention",
  supersedes: "replacement",
  "conflicts-with": "disagreement",
};
export interface KnowledgeRelation {
  id: string;
  from: string;
  to: string;
  type: RelationType;
}

export interface KnowledgeGraph {
  orgs: KnowledgeOrg[];
  projects: KnowledgeProject[];
  groups: KnowledgeGroup[];
  relations: KnowledgeRelation[];
}

/** Palette offered when creating a group, matching the app's status colours. */
export const GROUP_COLORS = [
  "var(--ac)",
  "var(--cyan)",
  "var(--green)",
  "var(--amber)",
  "var(--purple)",
  "var(--red)",
  "var(--tx3)",
];

const EMPTY_GRAPH: KnowledgeGraph = { orgs: [], projects: [], groups: [], relations: [] };

export const knowledge = {
  /** Every list is normalised: an older hostagent could answer null, and the
   *  screen maps over all four on first render. */
  async graph(): Promise<KnowledgeGraph> {
    const g = await call<Partial<KnowledgeGraph>>(HOST, "/knowledge");
    return {
      orgs: g.orgs ?? [],
      projects: g.projects ?? [],
      groups: (g.groups ?? []).map((x) => ({
        ...x,
        projects: x.projects ?? [],
        sources: x.sources ?? [],
      })),
      relations: g.relations ?? [],
    };
  },

  addOrg: (name: string) =>
    call<KnowledgeOrg>(HOST, "/knowledge/org", { method: "POST", body: JSON.stringify({ name }) }),
  renameOrg: (id: string, name: string) =>
    call<unknown>(HOST, "/knowledge/org", { method: "POST", body: JSON.stringify({ id, name }) }),
  deleteOrg: (id: string) =>
    call<unknown>(HOST, `/knowledge/org?id=${encodeURIComponent(id)}`, { method: "DELETE" }),

  addProject: (name: string, orgId: string) =>
    call<KnowledgeProject>(HOST, "/knowledge/project", {
      method: "POST",
      body: JSON.stringify({ name, orgId }),
    }),
  /** Moving replaces: a project belongs to exactly one organization. */
  moveProject: (id: string, orgId: string) =>
    call<unknown>(HOST, "/knowledge/project", { method: "POST", body: JSON.stringify({ id, orgId }) }),
  deleteProject: (id: string) =>
    call<unknown>(HOST, `/knowledge/project?id=${encodeURIComponent(id)}`, { method: "DELETE" }),

  addGroup: (name: string, color: string, owner = "", description = "") =>
    call<KnowledgeGroup>(HOST, "/knowledge/group", {
      method: "POST",
      body: JSON.stringify({ name, color, owner, description }),
    }),
  /** The id is not editable — it is the permission tag. */
  updateGroup: (g: Pick<KnowledgeGroup, "id" | "name" | "color" | "owner" | "description">) =>
    call<unknown>(HOST, "/knowledge/group", { method: "POST", body: JSON.stringify(g) }),
  deleteGroup: (id: string) =>
    call<unknown>(HOST, `/knowledge/group?id=${encodeURIComponent(id)}`, { method: "DELETE" }),

  /** Links are submitted as whole sets. Omitting a field leaves it alone;
   *  passing [] clears it — so never send [] for something you did not edit,
   *  or a group loses the sources it is allowed to retrieve. */
  setLinks: (groupId: string, links: { projects?: string[]; sources?: string[] }) =>
    call<unknown>(HOST, "/knowledge/group/links", {
      method: "POST",
      body: JSON.stringify({ groupId, ...links }),
    }),

  addRelation: (from: string, to: string, type: RelationType) =>
    call<KnowledgeRelation>(HOST, "/knowledge/relation", {
      method: "POST",
      body: JSON.stringify({ from, to, type }),
    }),
  setRelationType: (id: string, type: RelationType) =>
    call<unknown>(HOST, "/knowledge/relation", { method: "POST", body: JSON.stringify({ id, type }) }),
  deleteRelation: (id: string) =>
    call<unknown>(HOST, `/knowledge/relation?id=${encodeURIComponent(id)}`, { method: "DELETE" }),

  empty: () => EMPTY_GRAPH,
};

// --- the projected index ---

export interface GraphNeighbor {
  to: number;
  score: number;
}

export interface GraphNode {
  /** The identity: what group assignments store and what a search returns.
   *  For a local file this is the registered reference's id joined to the path
   *  within it, so two folders holding the same relative path stay two
   *  sources. */
  source: string;
  kind: "local" | "external";
  scope?: string;
  url?: string;
  /** Display name of the registered reference this came from — the folder an
   *  operator added. Older indexers omit it. */
  origin?: string;
  /** The path within that reference, which is the part a person recognises.
   *  Older indexers omit it; the source is then the whole of it. */
  rel?: string;
  groups?: string[];
  chunks: number;
  /** 0..1; the viewport decides the scale. */
  x: number;
  y: number;
  near: GraphNeighbor[];
}

export interface IndexGraph {
  nodes: GraphNode[];
  /** the projection collapsed — too few sources, or too little variation
   *  between them, to spread out. Say so rather than draw one dot. */
  degenerate: boolean;
}

// Pushing source→groups to the indexer used to live here. It does not any
// more: the host agent owns the graph and now pushes the mapping itself, at
// startup, after an edit, and before a run. Doing it from the front end made
// the permission model depend on somebody having the Knowledge screen open,
// and the indexer holds membership in memory — so a container restart, which
// registering a source deliberately causes, left it with no labels until the
// next time the screen was visited. See hostagent's knowledgesync.go.

export const indexGraph = {
  async load(): Promise<IndexGraph> {
    const g = await call<Partial<IndexGraph>>(RAG, "/graph");
    return {
      nodes: (g.nodes ?? []).map((n) => ({
        ...n,
        near: n.near ?? [],
        groups: n.groups ?? [],
        // An indexer from before sources were qualified sends neither, and its
        // paths are already the whole name — so the source stands in for both
        // and the screen groups everything under one heading.
        rel: n.rel || n.source,
        origin: n.origin ?? "",
      })),
      degenerate: g.degenerate ?? false,
    };
  },
};

/** shortLabel trims a path or URL to something that fits beside a node. */
export function shortLabel(source: string): string {
  return source.replace(/^https?:\/\//, "").split("/").slice(-2).join("/");
}

File: hostagent/internal/api/knowledge.go
package api

import (
	"net/http"
	"strings"

	"orchestra/hostagent/internal/store"
)

// The Knowledge graph's host-side API.
//
// The host owns this data: it is authored in the UI, must survive a restart,
// and must be editable while the indexer container is down. The indexer is fed
// from it rather than the other way round — same arrangement as knowledge
// sources themselves.
//
// Group ids are the permission tags the gateway hands to the indexer, so they
// are generated once at creation and never edited. Renaming a group changes
// only its label.

// knowledgeGraph is the whole graph in one response. The screen needs all of it
// on load — the rail lists groups, the canvas needs relations, the assignment
// view needs orgs and projects — so it is one round trip rather than four.
type knowledgeGraph struct {
	Orgs      []store.KnowledgeOrg      `json:"orgs"`
	Projects  []store.KnowledgeProject  `json:"projects"`
	Groups    []store.KnowledgeGroup    `json:"groups"`
	Relations []store.KnowledgeRelation `json:"relations"`
}

// storeReady answers with a 503 when persistence is unavailable, so the screen
// can say the graph cannot be loaded instead of rendering as if it were empty.
func (s *Server) storeReady(w http.ResponseWriter) bool {
	if s.store == nil {
		writeErr(w, 503, "knowledge storage is unavailable")
		return false
	}
	return true
}

func (s *Server) handleKnowledge(w http.ResponseWriter, _ *http.Request) {
	if !s.storeReady(w) {
		return
	}
	var g knowledgeGraph
	var err error
	if g.Orgs, err = s.store.KnowledgeOrgs(); err != nil {
		writeErr(w, 500, err.Error())
		return
	}
	if g.Projects, err = s.store.KnowledgeProjects(); err != nil {
		writeErr(w, 500, err.Error())
		return
	}
	if g.Groups, err = s.store.KnowledgeGroups(); err != nil {
		writeErr(w, 500, err.Error())
		return
	}
	if g.Relations, err = s.store.KnowledgeRelations(); err != nil {
		writeErr(w, 500, err.Error())
		return
	}
	writeJSON(w, 200, g)
}

// --- organizations ---

func (s *Server) handleKnowledgeOrg(w http.ResponseWriter, r *http.Request) {
	if !s.storeReady(w) {
		return
	}
	var req struct{ ID, Name string }
	if !decode(w, r, &req) {
		return
	}
	req.Name = strings.TrimSpace(req.Name)
	if req.Name == "" {
		writeErr(w, 400, "name is required")
		return
	}
	// An id in the body means "rename this one"; its absence means "create".
	if req.ID != "" {
		ok, err := s.store.RenameKnowledgeOrg(req.ID, req.Name)
		s.writeUpdated(w, ok, err, "organization")
		return
	}
	org, err := s.store.AddKnowledgeOrg(req.Name)
	if err != nil {
		writeErr(w, 500, err.Error())
		return
	}
	writeJSON(w, 201, org)
}

func (s *Server) handleKnowledgeOrgDelete(w http.ResponseWriter, r *http.Request) {
	if !s.storeReady(w) {
		return
	}
	ok, err := s.store.DeleteKnowledgeOrg(r.URL.Query().Get("id"))
	s.writeUpdated(w, ok, err, "organization")
}

// --- projects ---

func (s *Server) handleKnowledgeProject(w http.ResponseWriter, r *http.Request) {
	if !s.storeReady(w) {
		return
	}
	var req struct{ ID, Name, OrgID string }
	if !decode(w, r, &req) {
		return
	}
	req.Name = strings.TrimSpace(req.Name)
	if req.ID != "" {
		// Name and organization are edited independently, so an empty field
		// means "leave it alone" rather than "clear it".
		if req.Name != "" {
			if ok, err := s.store.RenameKnowledgeProject(req.ID, req.Name); err != nil || !ok {
				s.writeUpdated(w, ok, err, "project")
				return
			}
		}
		if req.OrgID != "" {
			if ok, err := s.store.MoveKnowledgeProject(req.ID, req.OrgID); err != nil || !ok {
				s.writeUpdated(w, ok, err, "project")
				return
			}
		}
		writeJSON(w, 200, map[string]string{"updated": req.ID})
		return
	}
	if req.Name == "" || req.OrgID == "" {
		writeErr(w, 400, "name and orgId are required")
		return
	}
	// Without an existence check the project would be created under an
	// organization that is not there, and would never appear in the tree.
	if ok, err := s.store.KnowledgeExists("knowledge_orgs", req.OrgID); err != nil {
		writeErr(w, 500, err.Error())
		return
	} else if !ok {
		writeErr(w, 404, "unknown organization: "+req.OrgID)
		return
	}
	p, err := s.store.AddKnowledgeProject(req.Name, req.OrgID)
	if err != nil {
		writeErr(w, 500, err.Error())
		return
	}
	writeJSON(w, 201, p)
}

func (s *Server) handleKnowledgeProjectDelete(w http.ResponseWriter, r *http.Request) {
	if !s.storeReady(w) {
		return
	}
	ok, err := s.store.DeleteKnowledgeProject(r.URL.Query().Get("id"))
	s.writeUpdated(w, ok, err, "project")
}

// --- groups ---

func (s *Server) handleKnowledgeGroup(w http.ResponseWriter, r *http.Request) {
	if !s.storeReady(w) {
		return
	}
	var req struct{ ID, Name, Color, Owner, Description string }
	if !decode(w, r, &req) {
		return
	}
	req.Name = strings.TrimSpace(req.Name)
	if req.Name == "" {
		writeErr(w, 400, "name is required")
		return
	}
	if req.ID != "" {
		ok, err := s.store.UpdateKnowledgeGroup(req.ID, req.Name, req.Color, req.Owner, req.Description)
		s.writeUpdated(w, ok, err, "group")
		return
	}
	g, err := s.store.AddKnowledgeGroup(req.Name, req.Color, req.Owner, req.Description)
	if err != nil {
		writeErr(w, 500, err.Error())
		return
	}
	writeJSON(w, 201, g)
}

func (s *Server) handleKnowledgeGroupDelete(w http.ResponseWriter, r *http.Request) {
	if !s.storeReady(w) {
		return
	}
	ok, err := s.store.DeleteKnowledgeGroup(r.URL.Query().Get("id"))
	// Deleting a group revokes every source it granted, which is the direction
	// worth being prompt about: the indexer should stop honouring it now, not
	// at the next run.
	if ok && err == nil {
		s.syncKnowledgeGroupsLogged("group deleted")
	}
	s.writeUpdated(w, ok, err, "group")
}

// handleKnowledgeGroupLinks replaces a group's project links or its source
// membership. Both are edited as a set in the UI, so both are submitted whole:
// sending a partial list would quietly revoke whatever was omitted, which for
// sources means revoking retrieval access.
func (s *Server) handleKnowledgeGroupLinks(w http.ResponseWriter, r *http.Request) {
	if !s.storeReady(w) {
		return
	}
	var req struct {
		GroupID  string   `json:"groupId"`
		Projects []string `json:"projects"`
		Sources  []string `json:"sources"`
	}
	if !decode(w, r, &req) {
		return
	}
	if req.GroupID == "" {
		writeErr(w, 400, "groupId is required")
		return
	}
	if ok, err := s.store.KnowledgeExists("knowledge_groups", req.GroupID); err != nil {
		writeErr(w, 500, err.Error())
		return
	} else if !ok {
		writeErr(w, 404, "unknown group: "+req.GroupID)
		return
	}
	// nil means "not submitted"; an empty slice means "clear it". Treating the
	// two alike would wipe a group's sources every time only its projects were
	// edited.
	if req.Projects != nil {
		if err := s.store.SetKnowledgeGroupProjects(req.GroupID, req.Projects); err != nil {
			writeErr(w, 500, err.Error())
			return
		}
	}
	if req.Sources != nil {
		if err := s.store.SetKnowledgeGroupSources(req.GroupID, req.Sources); err != nil {
			writeErr(w, 500, err.Error())
			return
		}
		// This is the edit that changes what the indexer may return, so it is
		// the one that is pushed. Project links do not: they decide which groups
		// a scope resolves to, which is answered here at launch, not there.
		s.syncKnowledgeGroupsLogged("group sources changed")
	}
	writeJSON(w, 200, map[string]string{"updated": req.GroupID})
}

// --- relations ---

func (s *Server) handleKnowledgeRelation(w http.ResponseWriter, r *http.Request) {
	if !s.storeReady(w) {
		return
	}
	var req struct{ ID, From, To, Type string }
	if !decode(w, r, &req) {
		return
	}
	if req.Type == "" || !RelationTypes[req.Type] {
		writeErr(w, 400, "unknown relation type: "+req.Type)
		return
	}
	if req.ID != "" {
		ok, err := s.store.SetKnowledgeRelationType(req.ID, req.Type)
		s.writeUpdated(w, ok, err, "relation")
		return
	}
	if req.From == "" || req.To == "" {
		writeErr(w, 400, "from and to are required")
		return
	}
	// A relation to itself draws nothing and expands to what the caller already
	// has, so it is a mistake rather than a degenerate case worth storing.
	if req.From == req.To {
		writeErr(w, 400, "a relation needs two distinct groups")
		return
	}
	for _, id := range []string{req.From, req.To} {
		if ok, err := s.store.KnowledgeExists("knowledge_groups", id); err != nil {
			writeErr(w, 500, err.Error())
			return
		} else if !ok {
			writeErr(w, 404, "unknown group: "+id)
			return
		}
	}
	rel, err := s.store.AddKnowledgeRelation(req.From, req.To, req.Type)
	if err != nil {
		writeErr(w, 500, err.Error())
		return
	}
	writeJSON(w, 201, rel)
}

func (s *Server) handleKnowledgeRelationDelete(w http.ResponseWriter, r *http.Request) {
	if !s.storeReady(w) {
		return
	}
	ok, err := s.store.DeleteKnowledgeRelation(r.URL.Query().Get("id"))
	s.writeUpdated(w, ok, err, "relation")
}

// writeUpdated turns a (found, err) pair into the one response shape these
// handlers share.
func (s *Server) writeUpdated(w http.ResponseWriter, ok bool, err error, what string) {
	switch {
	case err != nil:
		writeErr(w, 500, err.Error())
	case !ok:
		writeErr(w, 404, "unknown "+what)
	default:
		writeJSON(w, 200, map[string]string{"ok": what})
	}
}
